File: assignments/games-in-python/hangman.js
(() => {
  const words = [
    'hangman','python','challenge','variable','function','iteration','condition','module','package','developer'
  ];

  let secret = '';
  let correct = new Set();
  let wrong = new Set();
  let attempts = 6;

  const wordEl = document.getElementById('word');
  const wrongEl = document.getElementById('wrong');
  const attemptsEl = document.getElementById('attempts');
  const guessInput = document.getElementById('guess');
  const submitBtn = document.getElementById('submit');
  const newBtn = document.getElementById('new');
  const picEl = document.getElementById('hangman-pic');

  function choose() {
    return words[Math.floor(Math.random()*words.length)];
  }

  function render() {
    const revealed = secret.split('').map(c => correct.has(c) ? c : '_').join(' ');
    wordEl.textContent = revealed;
    wrongEl.textContent = wrong.size ? Array.from(wrong).sort().join(', ') : 'None';
    attemptsEl.textContent = attempts;
    picEl.textContent = '';
  }

  function checkWin() {
    if (secret.split('').every(c => correct.has(c))) {
      setTimeout(()=> alert(`You win! The word was '${secret}'.`), 10);
      return true;
    }
    if (attempts <= 0) {
      setTimeout(()=> alert(`Game over. The word was '${secret}'.`), 10);
      return true;
    }
    return false;
  }

  function start() {
    secret = choose();
    correct = new Set();
    wrong = new Set();
    attempts = 6;
    render();
  }

  function handleGuess() {
    const raw = guessInput.value.trim().toLowerCase();
    if (!raw) return;
    if (!/^[a-z]+$/.test(raw)) { alert('Please enter only letters.'); return; }

    if (raw.length === 1) {
      if (correct.has(raw) || wrong.has(raw)) { alert('Already tried.'); return; }
      if (secret.includes(raw)) {
        correct.add(raw);
      } else {
        wrong.add(raw);
        attempts -= 1;
      }
    } else {
      // full-word guess
      if (raw === secret) {
        secret.split('').forEach(c => correct.add(c));
      } else {
        attempts -= 1;
      }
    }

    guessInput.value = '';
    render();
    checkWin();
  }

  submitBtn.addEventListener('click', handleGuess);
  guessInput.addEventListener('keydown', (e)=> { if (e.key === 'Enter') handleGuess(); });
  newBtn.addEventListener('click', start);

  start();
})();

File: assignments/games-in-python/hangman.py
import random
import sys


HANGMAN_PICS = [
    "",
    "\n  O",
    "\n  O\n  |",
    "\n  O\n /|",
    "\n  O\n /|\\",
    "\n  O\n /|\\\n /",
    "\n  O\n /|\\\n / \\",
]


def choose_word():
    words = [
        "hangman",
        "python",
        "challenge",
        "variable",
        "function",
        "iteration",
        "condition",
        "module",
        "package",
        "developer",
    ]
    return random.choice(words).lower()


def display_state(secret, correct_letters, wrong_guesses, attempts_left):
    revealed = " ".join(c if c in correct_letters else "_" for c in secret)
    pic_index = len(HANGMAN_PICS) - 1 - attempts_left
    pic_index = max(0, min(pic_index, len(HANGMAN_PICS) - 1))
    print("\n" + HANGMAN_PICS[pic_index])
    print(f"Word: {revealed}")
    print(f"Wrong guesses: {', '.join(sorted(wrong_guesses)) if wrong_guesses else 'None'}")
    print(f"Attempts left: {attempts_left}")


def prompt_guess(used_letters):
    while True:
        guess = input("Guess a letter or the full word: ").strip().lower()
        if not guess:
            print("Please enter a letter or word.")
            continue
        if not guess.isalpha():
            print("Only alphabetic characters are allowed.")
            continue
        if len(guess) == 1 and guess in used_letters:
            print("You've already tried that letter.")
            continue
        return guess


def play_round():
    secret = choose_word()
    correct_letters = set()
    wrong_guesses = set()
    attempts_left = 6

    print("\n--- Hangman ---")

    while attempts_left > 0:
        display_state(secret, correct_letters, wrong_guesses, attempts_left)
        guess = prompt_guess(correct_letters | wrong_guesses)

        # full-word guess
        if len(guess) > 1:
            if guess == secret:
                print(f"Correct — the word was '{secret}'. You win!")
                return True
            else:
                attempts_left -= 1
                print("Incorrect full-word guess.")
                continue

        # single letter
        if guess in secret:
            correct_letters.add(guess)
            print(f"Good guess: '{guess}' is in the word.")
            if all(ch in correct_letters for ch in secret):
                print(f"You revealed the word '{secret}'. You win!")
                return True
        else:
            wrong_guesses.add(guess)
            attempts_left -= 1
            print(f"Nope: '{guess}' is not in the word.")

    print(f"\nNo attempts left. The word was '{secret}'. Game over.")
    return False


def main():
    print("Play Hangman — guess letters to reveal the word.")
    try:
        while True:
            play_round()
            again = input("Play again? (y/n): ").strip().lower()
            if not again or again[0] != 'y':
                print("Thanks for playing Hangman!")
                break
    except (KeyboardInterrupt, EOFError):
        print("\nGoodbye.")
        sys.exit(0)


if __name__ == '__main__':
    main()
